sys/usb/usbus/msc/scsi.c

```diff
@@ -0,0 +1,258 @@
+{
+    usbus_msc_device_t *msc = (usbus_msc_device_t*)handler;
+
+    if (cbw->data_len != 0) {
+        static const usbopt_enable_t enable = USBOPT_ENABLE;
+
+        if ((cbw->flags & USB_MSC_CBW_FLAG_IN) != 0) {
+            usbdev_ep_set(msc->ep_in->ep, USBOPT_EP_STALL, &enable,
+                          sizeof(usbopt_enable_t));
+        }
+        else {
+            usbdev_ep_set(msc->ep_out->ep, USBOPT_EP_STALL, &enable,
+                          sizeof(usbopt_enable_t));
+        }
+    }
+    msc->state = GEN_CSW;
+}
+
+void _scsi_write10(usbus_handler_t *handler, msc_cbw_buf_t *cbw)
+{
+    usbus_msc_device_t *msc = (usbus_msc_device_t*)handler;
+    msc_cbw_rw10_pkt_t *pkt = (msc_cbw_rw10_pkt_t *)cbw->cb;
+
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    msc->block = byteorder_swapl(pkt->blk_addr);
+
+    /* Get number of blocks to transfer */
+    msc->block_nb = byteorder_swaps(pkt->xfer_len);
+
+    msc->cmd.len = cbw->data_len;
+    msc->state = DATA_TRANSFER;
+}
+
+void _scsi_read10(usbus_handler_t *handler, msc_cbw_buf_t *cbw)
+{
+    usbus_msc_device_t *msc = (usbus_msc_device_t*)handler;
+    msc_cbw_rw10_pkt_t *pkt = (msc_cbw_rw10_pkt_t *)cbw->cb;
+
+    /* Convert big endian data from SCSI to little endian */
+    /* Get first block number to read from */
+    msc->block = byteorder_swapl(pkt->blk_addr);
+
+    /* Get number of blocks to transfer */
+    msc->block_nb = byteorder_swaps(pkt->xfer_len);
+
+    msc->cmd.len = cbw->data_len;
+    msc->state = DATA_TRANSFER;
+
+    if ((cbw->flags & USB_MSC_CBW_FLAG_IN) != 0) {
+            _rx_ready(msc);
+    }
+}
+
+void _scsi_inquiry(usbus_handler_t *handler)
+{
+    usbus_msc_device_t *msc = (usbus_msc_device_t*)handler;
+    msc_inquiry_pkt_t pkt;
+    size_t len = sizeof(msc_inquiry_pkt_t);
+    memset(&pkt, 0, len);
+    /* prepare pkt response */
+    pkt.type = 0;
+    pkt.removable = 1;
+    pkt.version = 0x01;
+    pkt.length = len - 4;
+    pkt.flags[0] = 0x80;
+
+    memcpy(&pkt.vendor_id, USBUS_MSC_VENDOR_ID, sizeof(pkt.vendor_id));
+    memcpy(&pkt.product_id, USBUS_MSC_PRODUCT_ID, sizeof(pkt.product_id));
+    memcpy(&pkt.product_rev, USBUS_MSC_PRODUCT_REV, sizeof(pkt.product_rev));
+
+    /* copy into ep buffer */
+    memcpy(msc->ep_in->ep->buf, &pkt, len);
+    usbdev_ep_ready(msc->ep_in->ep, len);
+    msc->state = WAIT_FOR_TRANSFER;
+    return;
+}
+
+void _scsi_read_capacity(usbus_handler_t *handler)
+{
+    usbus_msc_device_t *msc = (usbus_msc_device_t*)handler;
+    msc_read_capa_pkt_t pkt;
+    uint32_t blk_nb = (mtd0->sector_count * mtd0->pages_per_sector);
+    size_t len = sizeof(msc_read_capa_pkt_t);
+
+    assert(mtd0->page_size <= HOST_MINIMAL_PAGE_SIZE);
+
+    /* Host side expects to have at least 512 bytes per page, otherwise it will reject
+       the setup. Thus add some logic here to be able to use smaller page size mtd device */
```

Comment:
Can't we report a larger page size to the host?
I suppose then the statically allocated `buff` would be too small to act as a page cache.

(could we use `malloc()` here if we need the page cache at all? :innocent:)

Reply:
Why would you report a larger page size to the host ?

Reply:
Internal flash can have 4k or 8k pages.
If we want to support that too we would have to report larger pages - unless we also want to map that to virtual 512 byte pages.

Reply:
Reporting larger page size is somehow limited.
Linux rejects me when I was testing SAME54 flashpage because I reported 8192 bytes as page size.
So supported values are 512, 1024, 2048. (maybe 4096 I didn't try).